src/lib/Transform.ts

```diff
@@ -134,52 +140,76 @@ export function batch<T>(
 	let writeBatchPromise = Promise.resolve();
 	let timer: NodeJS.Timer | undefined;
 
+	async function queueNextWrite(peeled: T[]) {
+		await writeBatchPromise;
+		if (!isAborted) {
+			await writable.write(peeled);
+		}
+	}
+
 	async function flush() {
-		if (timer) {
+		if (timer !== undefined) {
 			clearTimeout(timer);
+			timer = undefined;
 		}
 
 		if (queue.length) {
 			const peeled = queue;
 			queue = [];
 
-			writeBatchPromise = writeBatchPromise.then(() =>
-				isAborted ? Promise.resolve() : writable.write(peeled)
-			);
-
-			return writeBatchPromise;
-		} else {
-			return writeBatchPromise;
+			writeBatchPromise = queueNextWrite(peeled);
 		}
+
+		await writeBatchPromise;
 	}
 
-	function cleanup() {
-		if (timer) {
-			clearTimeout(timer);
+	let outOfFlowFlushPromise: TrackedVoidPromise | undefined;
```

Comment:
I find this logic pretty complex, and am looking for ways to simplify it.

I'm thinking:
- Some writes come in, at a certain time they become bigger than minBatchSize, so you start writing them.
- You flip a bool: `writing = true;`
- You await that write, to intercept any possible error, in which case you'll abort the stream (we can see if we can even prevent that, too), and to set `writing = false` in any case.
- You indeed don't return this promise to upstream, such that more writes keep coming in, and the emptied queue starts filling up again
- Once the queue again reaches `minBatchSize`, you basically want to ignore it, because the previous one is still pending (`writing === true`)
- Once the queue then reaches `maxBatchSize` you can simply `write()` that whole batch (irrespective of the state of `writing`), because `Stream` will guarantee the queueing of the writes itself already. You DO return this promise to upstream, such that even further incoming writes are blocked until both of the previous writes have finished again.

You'll need another check: because you then ignore writes after `minBatchSize` is reached again, after the pending `writing` is finished and set back to false, you may need to start another write immediately if the queue already reached `minBatchSize` in the mean time.

I think you could then even prevent aborting, if instead of storing the fact that you're writing in a bool, you store the actual promise for that write in a variable, or reset that variable back to `undefined` once you passed it upstream. You can then always just `const p = pendingWrite; pendingWrite = undefined; await p;` before you start the `maxBatchSize` flush, which will bubble that stored exception back up.

So, both types of 'partial flush' (i.e. `minBatchSize` and the one triggered on timeout) should only trigger when `pendingWrite !== undefined`, and should then set `pendingWrite` to the promise returned from the downstream write.

Reply:
Armed with your excellent tests, I thought I'd give this a spin.

If we want to just abort the stream whenever a `minBatchSize` write fails, the following seems to be sufficient:

```ts
export function batch<T>(
	readable: Readable<T>,
	writable: Writable<T[]>,
	maxBatchSize: number,
	minBatchSize = maxBatchSize,
	flushTimeout: number | undefined
): void {
	let isAborted = false;

	writable.aborted().catch((err: Error) => {
		readable.abort(err);
		isAborted = true;
	});
	readable.aborted().catch((err) => {
		writable.abort(err);
	});

	let queue: T[] = [];
	let timer: NodeJS.Timer | undefined;
	let pendingWrite = false;

	async function flush() {
		if (timer !== undefined) {
			clearTimeout(timer);
			timer = undefined;
		}

		if (queue.length) {
			const peeled = queue;
			queue = [];

			if (!isAborted) {
				await writable.write(peeled);
			}
		}
	}

	async function earlyFlush(): Promise<void> {
		try {
			do {
				pendingWrite = true;
				await flush();
				pendingWrite = false; // doesn't matter if not reset on error
			} while (queue.length >= minBatchSize);
		} catch (err) {
			readable.abort(err);
		}
	}

	readable.forEach(
		async (v: T): Promise<void> => {
			queue.push(v);

			if (queue.length >= maxBatchSize) {
				// backpressure
				await flush();
			} else if (queue.length >= minBatchSize && !pendingWrite) {
				// no backpressure yet (until new queue fills to maxBatchSize)
				earlyFlush();
			}

			if (queue.length && flushTimeout !== undefined && !pendingWrite) {
				if (timer !== undefined) {
					clearTimeout(timer);
				}
				timer = setTimeout(earlyFlush, flushTimeout);
			}
		},
		async (error?: Error) => {
			try {
				await flush();
				await writable.end(error, readable.result());
			} catch (e) {
				await writable.end(e, readable.result());
			}
		},
		flush
	);
}
```

This passes all of your tests! :tada: 

I tried adding an explicit test case for that while loop in earlyFlush (without it the tests also pass, but I think that's because we're missing a test case :)). It's getting time to sleep though.

To do the 'remember the error and return it on the next possible incoming write'-thing appears to be a bit more tricky though...

The algorithm I proposed in my previous post has an issue that it could 'delay' giving back an error from a `minBatchSize` write until the next `maxBatchSize` write would happen, which is undesirable. So we'd need to always return that `pendingWrite` promise from an incoming write whenever that `pendingWrite` has errored already. And whenever we did somehow return it to the write, we'd need to make sure we no longer return or otherwise directly await it anymore, to prevent such an error from being 'bounced back' twice.

Also, if there was such an error, this would mean the pending write has basically completed, but we'd still want to already start the next batch in the mean time. Otherwise, if no further write would come in 'soon', we'd never actually start that next batch (until the stream is ended).

I totally understand if you don't want to spend the effort try to figure this out, btw. In that case, I may give it another go later this week so see if there's a clean way to do it. Otherwise, I may want to change the `maxBatchSize` write to _also_ abort the stream in case of an error, because I think it'd be a bit inconsistent to sometimes return the error, and sometimes abort the stream.

Reply:
This looks great! I've added it. Thank you!

Reply:
> I tried adding an explicit test case for that while loop in earlyFlush

I've now added a test for this case.

Reply:
Added the functionality to retain errors and throw them at next read or end. It seems to work. Requesting feedback on a couple of points:

- [ ] Any cases I missed around errors?
- [ ] Is the ender written correctly? Not just in terms of behaviour but idiomatically. There have been so many changes I just don't know anymore.
- [ ] Issue from an earlier comment: should a failed batch be returned to the internal queue?

Reply:
- Any cases I missed around errors?
  - Don't think so, will look it over once more after error throwing is moved 'down'. I think the solution you now have is elegant (esp. with that function to 'consume' the error).
- Is the ender written correctly? Not just in terms of behaviour but idiomatically. There have been so many changes I just don't know anymore.
  - Addressed this in a review comment.
- Issue from an earlier comment: should a failed batch be returned to the internal queue?
  - Every write must be somehow handled, either error returned, or processed successfully. There's a bit of gray area here in that if there's an early flush, you may receive that error on a later write, so it will be hard for a source to know exactly which elements were processed or not. But I suppose that's than just inevitable when someone decides to use `minBatchSize` smaller than `maxBatchSize`. Perhaps a note in the docs of `batch()` would be good.